Wobble/Scheduling/TaskHandler.cs

```diff
@@ -71,42 +82,56 @@ public CancellationToken Run(T input, int delay = 0)
                 Source.Cancel();
                 Source.Dispose();
                 Source = new CancellationTokenSource();
-
-                IsCompleted = false;
-                IsCancelled = false;
-                IsRunning = true;
+                TaskRun(Source.Token, input, delay);
             }
 
-            var token = Source.Token;
-            Task.Run(async () =>
+            return Source.Token;
+        }
+
+        /// <summary>
+        ///     This method will run the task asynchronously.
+        /// </summary>
+        /// <param name="token"></param>
+        /// <param name="input"></param>
+        /// <param name="delay"></param>
+        private async void TaskRun(CancellationToken token, T input, int delay = 0)
+        {
+            try
             {
+                token.ThrowIfCancellationRequested();
+                await SemaphoreSlim.WaitAsync(token);
+
                 try
                 {
-                    await Task.Delay(delay, token);
+                    IsCompleted = false;
```

Comment:
~~Since you're protecting these variables with a semaphore, you must change `Cancel()` and `Dispose()` to using the semaphore, too. I think they should be using the semaphore _in addition_ to the already-present `lock {}` (inside of it) to guarantee no invalid code paths, but I'm not sure.~~

You actually still need to use `lock {}` around these variables to synchronize properly with `Cancel()` and `Dispose()`. This is because you cannot actually add semaphore synchronization into `Cancel()` and `Dispose()` because that would block them until the task finishes (since the semaphore is held for the whole duration of the task), which is undesirable.

To check for the data races I suspect can happen, try running the task and cancelling it from a different thread many times, and observing if these state variables are set correctly. In particular, with the current implementation it's possible to `Cancel()` the task, for example, after `IsCancelled = false;` but before `IsRunning = true;` which means that the task will be running with `IsCancelled = true;` _and_ `IsRunning = true;`.

I think this would bring it considerably back to the previous design, so I suggest fixing the deadlock in the previous design rather than using this new design.

Reply:
Hmm, actually for that to trigger you'd need at least two threads running the task so it can be stuck on the semaphore... Currently trying to make the race.

Reply:
Yep, got a race: https://repl.it/repls/RunnyInsecureBucket

I was looking for that invalid state, but I actually got a crash from trying to use the token after it has been disposed of in a different thread. Might take a couple of runs to reproduce.

Reply:
That makes sense. Wouldn't it be more appropriate to use the semaphore in the `Cancel()` and `Dispose()` methods?

The issue with the previous design was not being able the properly synchronize the `task.delay()`

Reply:
It _would prevent data races_, but it would make `Cancel()` and `Dispose()` block until the currently running task finishes, which is not supposed to happen.

Reply:
I could take a look at the previous design again and see if I can fix the deadlock there without significant changes.

Reply:
Or wouldn't it make sense to apply `IsCancelled = true` in the operation canceled block?

Reply:
> It _would prevent data races_, but it would make `Cancel()` and `Dispose()` block until the currently running task finishes, which is not supposed to happen.

Since we're passing the cancellation token in every async method call, shouldn't the tasks cancel almost immediately after `Cancel()` is called?

Reply:
> Since we're passing the cancellation token in every async method call, shouldn't the tasks cancel almost immediately after Cancel() is called?

If they do IO and use it, then yes, they should. If they don't (like in the visual test for this :)) then it'll block.

Reply:
It turns out that the previous tasks do actually cancel immediately instead of the current task waiting for the current task to cancel.

```cs
// Initial Input
            var input = RNG.Next(1000, 9999);
            Console.WriteLine($"INITIAL INPUT: {input}");

            // This Code will Run Random Numbers Task 10 times at once. Previous tasks will automatically be cancelled.
            for (var i = 0; i < 10; i++)
            {
                RandomNumbersTask.Run(input, 2000);
                input = RNG.Next(1000, 9999);
            }

            // The previous tasks should've been cancelled and the output in the current task should match the final input.
            Console.WriteLine($"FINAL INPUT: {input}");
            RandomNumbersTask.Run(input, 2000);
```